code/network/multi_interpolate.cpp

```diff
@@ -0,0 +1,334 @@
+
+			aggregate_time = (Multi_Timing_Info.is_most_recent_frame(frame, player_index)) ? Multi_Timing_Info.get_aggregate_time(player_index) : Multi_Timing_Info.get_previous_aggregate_time(frame, player_index);
+
+			// if this is now the most recent packet, then make sure the next round of interpolation uses its position, even if the packet came in too late.
+			if (packet == _packets.begin()) {
+				_packets_expended = false;
+			}
+
+			// yes, insert is somewhat inefficient here, but it *greatly* simplifies the code
+			_packets.insert(packet, packet_info::packet_info(frame, aggregate_time, position, velocity, rotational_velocity, desired_velocity, desired_rotational_velocity, angles));
+			
+			// when the size of vector is 2, then we have just received enough pakets to begin interpolating,
+			// and we need to mark this object as ready to interpolate
+			if (_packets.size() == 2) {
+				_upcoming_packet_index = 0;
+				_prev_packet_index = 1;
+			}
+
+			if (static_cast<int>(_packets.size()) > PACKET_INFO_LIMIT) {
+				_packets.pop_back();
+			}
+
+			return;
+		}
+	}
+
+
+}
+
+// basically, copy the state from the object that had been simulated up to that point and treat it as the "old packet".
+// should never replace index 0
+void interpolation_manager::replace_packet(int index, vec3d* pos, matrix* orient, physics_info* pip) 
+{
+	// the hackiest part of the hack? Setting its frame. Let FSO think that it was basically brand new.
+	// it needs to handle it this way because otherwise another packet might get placed in front of it, 
+	// and we lose our intended effect of interpolating the simulation error away.
+	_packets[index].frame = _packets[index - 1].frame - 1;
+
+	_packets[index].aggregate_time = Multi_Timing_Info.get_last_time();
+	_packets[index].position = *pos;
+	_packets[index].velocity = pip->vel;
+	_packets[index].desired_velocity = pip->desired_vel;
+	_packets[index].rotational_velocity = pip->rotvel;
+	_packets[index].desired_rotational_velocity = pip->desired_rotvel; 
+
+	vm_extract_angles_matrix(&_packets[index].angle, orient);
+}
+
+/////////////////////////////////
+// Time Records Manager functions
+
+// run every 100 local frames to prune. We're never going to get these frame times anyway.  
+void multiplayer_timing_info::prune_frame_status(int player_index)
+{
+	int test_value = _most_recent_frames[player_index].front().first - 100;
+	SCP_vector<int> to_remove;
+
+	// find the bad entries
+	for (auto& entry : _frame_status[player_index]){
+		if (entry.second < test_value){
+			to_remove.push_back(entry.first);
+		}
+	}
+
+	// remove them
+	for (auto& culled_item : to_remove) {
+		_frame_status[player_index].erase(culled_item);
+	}
+}
+
+// make sure that we recalculate the current local time
+void multiplayer_timing_info::update_current_time() {
+	_current_time = f2i(static_cast<std::int64_t>(Missiontime) * TIMESTAMP_FREQUENCY) - _start_time + _skipped_time;
```

Comment:
This conversion doesn't produce anything, really.  Take a look at `timestamp_get_mission_time()` which is how `Missiontime` is calculated.

Incidentally, can't you just work with `Missiontime` directly, without converting it to anything?

Reply:
I see what you're saying now.  f2i takes a fix, so it wasn't keeping the 32 bit word size.  It was to prevent overflow, so maybe it will properly do that now that I just added the bitshift.

The time information from the other machine are coming in as timestamp deltas.  Converting `Missiontime` here allows us to work in the same format as the incoming data.  I could change what it sends, but that would require a multi version bump.

Reply:
Oh, I misinterpreted your previous code as referring to the timer frequency, not timestamp frequency.  Please rebase onto latest master and use `MILLISECONDS_PER_SECOND` instead.

I see what you mean about converting the time to milliseconds.  Though an extra comment to indicate the units would be helpful.

Most of the other code that works with `Missiontime` converts the deltas to `fix`.  For example:
```
		//	If a very long time since attacked, attack no matter what!
		if (Missiontime - aip->last_attack_time > i2f(6)) {
```
Is that an option?

Reply:
I changed everything over to timestamp, because something seemed to be going wrong with the calculation.  Later investigations point to it being something else, but I started using a basic timestamp() call until it's more stable (it's no longer crashing, but we have to make the movements a lot smoother now)

Reply:
Ok.